src/dht.cpp

```diff
@@ -1341,41 +1398,11 @@ Dht::searchStep(std::shared_ptr<Search> sr)
         while (sent and sr->currentlySolicitedNodeCount() < MAX_REQUESTED_SEARCH_NODES);
         /*DHT_LOG.DEBUG("[search %s IPv%c] step: sent %u requests (total %u).",
             sr->id.toString().c_str(), sr->af == AF_INET ? '4' : '6', i, sr->currentlySolicitedNodeCount());*/
+    }
 
-        auto expiredn = (size_t)std::count_if(sr->nodes.begin(), sr->nodes.end(), [&](const SearchNode& sn) {
-                    return sn.candidate or sn.node->isExpired();
-                });
-        if (i == 0 && expiredn == sr->nodes.size())
-        {
-            DHT_LOG.WARN("[search %s IPv%c] expired", sr->id.toString().c_str(), sr->af == AF_INET ? '4' : '6');
-            // no nodes or all expired nodes
-            sr->expired = true;
-            if (sr->announce.empty() && sr->listeners.empty()) {
-                // Listening or announcing requires keeping the cluster up to date.
-                sr->setDone();
-            }
-            {
-                auto get_cbs = std::move(sr->callbacks);
-                for (const auto& g : get_cbs) {
-                    if (g.second.done_cb)
-                        g.second.done_cb(false, {});
-                }
-            }
-            {
-                std::vector<DoneCallback> a_cbs;
-                a_cbs.reserve(sr->announce.size());
-                for (auto ait = sr->announce.begin() ; ait != sr->announce.end(); ) {
-                    if (ait->callback)
-                        a_cbs.emplace_back(std::move(ait->callback));
-                    if (not ait->permanent)
-                        ait = sr->announce.erase(ait);
-                    else
-                        ait++;
-                }
-                for (const auto& a : a_cbs)
-                    a(false, {});
-            }
-        }
+    if (sr->getNumberOfBadNodes() >= std::min(sr->nodes.size(), static_cast<size_t>(SEARCH_MAX_BAD_NODES))) {
```

Comment:
> BTW we should use that value to compute the number of nodes instead of 0.5 since it's the worst case scenario to support !

The number 25 is based on table 3 from the annex 1 of #137 (k=3/4). Going from 3/4 to 7/8, the number is at least 50. Which seems odd to me.

> Search expiration should happen only when we believe that we are disconnected from the network, if there is no unexpired nodes left in the search (or if the first N nodes are expired since we try them in order), otherwise we should aim to remember bad nodes as much as possible to avoid trying them again and block the search.

If I undestand correctly, you suggest to exclusively consider **expired** nodes in the node count, right? I also don't count consecutive expired nodes. May be that would be more appropriate.

- [ ] count consecutive **expired** nodes;

> The current method to check for expiration is probably well suited for our purpose, we only need to limit the maximum size of the search, and it will expire when every of its node is.

I'm not sure I get what you're saying. My understanding is that this patch should remove expired nodes from the affected `Search` before we reach the point where **all** nodes are expired. Hence the fact I replaced the expiration criteria to consider `min(nodes.size(), SEARCH_MAX_BAD_NODES)`. However, as you stated above, I should only consider expired nodes exclusively.